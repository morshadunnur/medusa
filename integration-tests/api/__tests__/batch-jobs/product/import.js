const path = require("path")

const setupServer = require("../../../../helpers/setup-server")
const { useApi } = require("../../../../helpers/use-api")
const { initDb, useDb } = require("../../../../helpers/use-db")

const adminSeeder = require("../../../helpers/admin-seeder")
const batchJobSeeder = require("../../../helpers/batch-job-seeder")
const userSeeder = require("../../../helpers/user-seeder")

const adminReqConfig = {
  headers: {
    Authorization: "Bearer test_token",
  },
}

jest.setTimeout(1000000)

describe("Product import batch job", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", "..", ".."))
    dbConnection = await initDb({ cwd })

    medusaProcess = await setupServer({
      cwd,
      redisUrl: "redis://127.0.0.1:6379",
      uploadDir: __dirname,
      verbose: true,
    })
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  beforeEach(async () => {
    try {
      await batchJobSeeder(dbConnection)
      await adminSeeder(dbConnection)
      await userSeeder(dbConnection)
    } catch (e) {
      console.log(e)
      throw e
    }
  })

  afterEach(async () => {
    const db = useDb()
    await db.teardown()
  })

  it("should import a csv file", async () => {
    jest.setTimeout(1000000)
    const api = useApi()

    const response = await api.post(
      "/admin/batch-jobs",
      {
        type: "product_import",
        context: {
          fileKey: "product-import.csv",
        },
      },
      adminReqConfig
    )

    console.log(response.data)

    const batchJobId = response.data.batch_job.id

    expect(batchJobId).toBeTruthy()

    // Pull to check the status until it is completed
    let batchJob
    let shouldContinuePulling = true
    while (shouldContinuePulling) {
      const res = await api.get(
        `/admin/batch-jobs/${batchJobId}`,
        adminReqConfig
      )

      await new Promise((resolve, _) => {
        setTimeout(resolve, 1000)
      })

      batchJob = res.data.batch_job

      // TODO: check, sometime the job is randomly stuck in a state (mostly "created" but sometimes in "processing" also)
      shouldContinuePulling = !(
        batchJob.status === "completed" || batchJob.status === "failed"
      )
    }

    expect(batchJob.status).toBe("completed")

    const productsResponse = await api.get("/admin/products", adminReqConfig)

    expect(productsResponse.data.count).toBe(2)
    expect(productsResponse.data.products).toMatchSnapshot()
  })
})
