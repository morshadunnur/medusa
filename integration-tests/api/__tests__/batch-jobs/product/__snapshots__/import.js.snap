// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product import batch job should import a csv file 1`] = `
Array [
  Object {
    "collection": null,
    "collection_id": null,
    "created_at": "2022-06-27T18:33:05.132Z",
    "deleted_at": null,
    "description": "test-product-description-1",
    "discountable": true,
    "external_id": null,
    "handle": "test-product-product-1",
    "height": null,
    "hs_code": null,
    "id": "O6S1YQ6mKm",
    "images": Array [
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "img_01G6K70HVG7FKHAKQFP6D8K9GQ",
        "metadata": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "url": "test-image.png",
      },
    ],
    "is_giftcard": false,
    "length": null,
    "material": null,
    "metadata": null,
    "mid_code": null,
    "options": Array [
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "opt_01G6K70HWA5Q2NBVWPTBZP2S51",
        "metadata": null,
        "product_id": "O6S1YQ6mKm",
        "title": "test-option-1",
        "updated_at": "2022-06-27T18:33:05.132Z",
      },
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "opt_01G6K70HWB0E8Z56CJC4G15QHN",
        "metadata": null,
        "product_id": "O6S1YQ6mKm",
        "title": "test-option-2",
        "updated_at": "2022-06-27T18:33:05.132Z",
      },
    ],
    "origin_country": null,
    "profile_id": "sp_01G6K70C0Q7S2W4DRZWYJS7AQM",
    "status": "draft",
    "subtitle": null,
    "tags": Array [
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "ptag_01G6K70HVP7XCKF6F3NSZYB0MX",
        "metadata": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "value": "123_1",
      },
    ],
    "thumbnail": "test-image.png",
    "title": "Test product",
    "type": null,
    "type_id": null,
    "updated_at": "2022-06-27T18:33:05.132Z",
    "variants": Array [
      Object {
        "allow_backorder": false,
        "barcode": "test-barcode-1",
        "calculated_price": null,
        "calculated_price_incl_tax": null,
        "calculated_tax": null,
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "ean": null,
        "height": null,
        "hs_code": null,
        "id": "variant_01G6K70HXDSQSBAD2HFJ60JK35",
        "inventory_quantity": 10,
        "length": null,
        "manage_inventory": true,
        "material": null,
        "metadata": null,
        "mid_code": null,
        "options": Array [
          Object {
            "created_at": "2022-06-27T18:33:05.132Z",
            "deleted_at": null,
            "id": "optval_01G6K70HXDDD3QV2EZYP85GH79",
            "metadata": null,
            "option_id": "opt_01G6K70HWA5Q2NBVWPTBZP2S51",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "value": "option 1 value red",
            "variant_id": "variant_01G6K70HXDSQSBAD2HFJ60JK35",
          },
          Object {
            "created_at": "2022-06-27T18:33:05.132Z",
            "deleted_at": null,
            "id": "optval_01G6K70HXDK5YJS1CB4EHSXF9V",
            "metadata": null,
            "option_id": "opt_01G6K70HWB0E8Z56CJC4G15QHN",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "value": "option 2 value 1",
            "variant_id": "variant_01G6K70HXDSQSBAD2HFJ60JK35",
          },
        ],
        "origin_country": null,
        "original_price": null,
        "original_price_incl_tax": null,
        "original_tax": null,
        "prices": Array [
          Object {
            "amount": 100,
            "created_at": "2022-06-27T18:33:05.132Z",
            "currency_code": "eur",
            "deleted_at": null,
            "id": "ma_01G6K70HXXQN8512V5SWTGBX3C",
            "max_quantity": null,
            "min_quantity": null,
            "price_list": null,
            "price_list_id": null,
            "region_id": "region-product-import-0",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "variant_id": "variant_01G6K70HXDSQSBAD2HFJ60JK35",
          },
          Object {
            "amount": 110,
            "created_at": "2022-06-27T18:33:05.132Z",
            "currency_code": "usd",
            "deleted_at": null,
            "id": "ma_01G6K70HY0J8GVXPQXX78VJRT7",
            "max_quantity": null,
            "min_quantity": null,
            "price_list": null,
            "price_list_id": null,
            "region_id": null,
            "updated_at": "2022-06-27T18:33:05.132Z",
            "variant_id": "variant_01G6K70HXDSQSBAD2HFJ60JK35",
          },
          Object {
            "amount": 130,
            "created_at": "2022-06-27T18:33:05.132Z",
            "currency_code": "dkk",
            "deleted_at": null,
            "id": "ma_01G6K70HY40F43M4R6V0T58ZG9",
            "max_quantity": null,
            "min_quantity": null,
            "price_list": null,
            "price_list_id": null,
            "region_id": "region-product-import-1",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "variant_id": "variant_01G6K70HXDSQSBAD2HFJ60JK35",
          },
        ],
        "product_id": "O6S1YQ6mKm",
        "sku": "test-sku-1",
        "tax_rates": null,
        "title": "Test variant",
        "upc": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "weight": null,
        "width": null,
      },
    ],
    "weight": null,
    "width": null,
  },
  Object {
    "collection": null,
    "collection_id": null,
    "created_at": "2022-06-27T18:33:05.132Z",
    "deleted_at": null,
    "description": "test-product-description",
    "discountable": true,
    "external_id": null,
    "handle": "test-product-product-2",
    "height": null,
    "hs_code": null,
    "id": "5VxiEkmnPV",
    "images": Array [
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "img_01G6K70HVG7FKHAKQFP6D8K9GQ",
        "metadata": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "url": "test-image.png",
      },
    ],
    "is_giftcard": false,
    "length": null,
    "material": null,
    "metadata": null,
    "mid_code": null,
    "options": Array [
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "opt_01G6K70HWTCFR2AFYFYPGHEMBJ",
        "metadata": null,
        "product_id": "5VxiEkmnPV",
        "title": "test-option",
        "updated_at": "2022-06-27T18:33:05.132Z",
      },
    ],
    "origin_country": null,
    "profile_id": "sp_01G6K70C0Q7S2W4DRZWYJS7AQM",
    "status": "draft",
    "subtitle": null,
    "tags": Array [
      Object {
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "id": "ptag_01G6K70HWN7VYDZA79PFJAK9ZA",
        "metadata": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "value": "123",
      },
    ],
    "thumbnail": "test-image.png",
    "title": "Test product",
    "type": null,
    "type_id": null,
    "updated_at": "2022-06-27T18:33:05.132Z",
    "variants": Array [
      Object {
        "allow_backorder": false,
        "barcode": "test-barcode-2",
        "calculated_price": null,
        "calculated_price_incl_tax": null,
        "calculated_tax": null,
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "ean": null,
        "height": null,
        "hs_code": null,
        "id": "variant_01G6K70HYAJW9DG7NYC1ECYKGA",
        "inventory_quantity": 10,
        "length": null,
        "manage_inventory": true,
        "material": null,
        "metadata": null,
        "mid_code": null,
        "options": Array [
          Object {
            "created_at": "2022-06-27T18:33:05.132Z",
            "deleted_at": null,
            "id": "optval_01G6K70HYB4CB1Q2R1K4Z3AAZE",
            "metadata": null,
            "option_id": "opt_01G6K70HWTCFR2AFYFYPGHEMBJ",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "value": "Option 1 value 1",
            "variant_id": "variant_01G6K70HYAJW9DG7NYC1ECYKGA",
          },
        ],
        "origin_country": null,
        "original_price": null,
        "original_price_incl_tax": null,
        "original_tax": null,
        "prices": Array [
          Object {
            "amount": 110,
            "created_at": "2022-06-27T18:33:05.132Z",
            "currency_code": "dkk",
            "deleted_at": null,
            "id": "ma_01G6K70HYN4PYY62PW12EYBH75",
            "max_quantity": null,
            "min_quantity": null,
            "price_list": null,
            "price_list_id": null,
            "region_id": "region-product-import-2",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "variant_id": "variant_01G6K70HYAJW9DG7NYC1ECYKGA",
          },
        ],
        "product_id": "5VxiEkmnPV",
        "sku": "test-sku-2",
        "tax_rates": null,
        "title": "Test variant",
        "upc": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "weight": null,
        "width": null,
      },
      Object {
        "allow_backorder": false,
        "barcode": "test-barcode-3",
        "calculated_price": null,
        "calculated_price_incl_tax": null,
        "calculated_tax": null,
        "created_at": "2022-06-27T18:33:05.132Z",
        "deleted_at": null,
        "ean": null,
        "height": null,
        "hs_code": null,
        "id": "variant_01G6K70HZ0V5DTA47GDAV9FV9D",
        "inventory_quantity": 10,
        "length": null,
        "manage_inventory": true,
        "material": null,
        "metadata": null,
        "mid_code": null,
        "options": Array [
          Object {
            "created_at": "2022-06-27T18:33:05.132Z",
            "deleted_at": null,
            "id": "optval_01G6K70HZ052XAGZR1A5C0KX97",
            "metadata": null,
            "option_id": "opt_01G6K70HWTCFR2AFYFYPGHEMBJ",
            "updated_at": "2022-06-27T18:33:05.132Z",
            "value": "Option 1 Value blue",
            "variant_id": "variant_01G6K70HZ0V5DTA47GDAV9FV9D",
          },
        ],
        "origin_country": null,
        "original_price": null,
        "original_price_incl_tax": null,
        "original_tax": null,
        "prices": Array [
          Object {
            "amount": 120,
            "created_at": "2022-06-27T18:33:05.132Z",
            "currency_code": "usd",
            "deleted_at": null,
            "id": "ma_01G6K70HZ1MEAQEKNV8ZG6E3HW",
            "max_quantity": null,
            "min_quantity": null,
            "price_list": null,
            "price_list_id": null,
            "region_id": null,
            "updated_at": "2022-06-27T18:33:05.132Z",
            "variant_id": "variant_01G6K70HZ0V5DTA47GDAV9FV9D",
          },
        ],
        "product_id": "5VxiEkmnPV",
        "sku": "test-sku-3",
        "tax_rates": null,
        "title": "Test variant",
        "upc": null,
        "updated_at": "2022-06-27T18:33:05.132Z",
        "weight": null,
        "width": null,
      },
    ],
    "weight": null,
    "width": null,
  },
]
`;
